### DISCUSSION TITLE
Binding Inputs to Relationship Properties

### DISCUSSION THREAD

**Discussion by benjibee** [livewire/livewire#5090] [7. Bugs (archived)] [answered]
I have a `rentalCar` entity that has a relationship with `rentalCarReturnFee` entities.
```php
    public RentalCar $rentalCar;

    protected $rules = [
        'rentalCar.rentalCarReturnFees.*.fee' => ['required', 'int'],
        'rentalCar.rentalCarReturnFees.*.pick_up_location_id' => ['required'],
        'rentalCar.rentalCarReturnFees.*.drop_off_location_id' => ['required'],
    ];
```

Attempting to bind inputs in the Blade file is not working as expected.

```blade
@foreach($rentalCar->rentalCarReturnFees as $index => $rentalCarReturnFee)

    {{-- Works --}}
    @dump($rentalCar->rentalCarReturnFees[$index]->fee)

    {{-- Doesn't work --}}
     <input type="text" wire:model.lazy="rentalCar.rentalCarReturnFees.{{ $index }}.fee">
```

When interacting with the elements I see this behavior, which may be a very clear sign of what's going wrong. I just don't know it.

https://user-images.githubusercontent.com/620119/177609170-39798b3b-c491-40e0-b6b8-6be7abecd201.mov

I could provide all the context and more code, I'm sure, but as the `@dump` is working perfect within the `@foreach` loop it's clear that the rest of the data is being handled correctly. What am I missing?

**Comment by benjibee** [comment]
So after some digging I think I've found the source of the issue! It's not at all what I expected but it looks like Laravel's `toArray()` method is converting the model of the relationship from camel case to snake case.

With that hint I was able to find #3803, though not solid workaround.

If I convert the name of the relationship's model (not the parent model itself) to snake case in both the rules and the Blade file the data is bound to the inputs but throws an error upon attempting to save the data:

```php
  // Doesn't work.
  'rentalCar.rentalCarReturnFees.*.fee' => ['required', 'int'],

  // Kinda works.
  'rentalCar.rental_car_return_fees.*.fee' => ['required', 'int'],
```

```blade
  // Doesn't work.
  <input type="text" wire:model.lazy="rentalCar.rentalCarReturnFees.{{ $index }}.fee">

  // Kinda works.
  <input type="text" wire:model.lazy="rentalCar.rental_car_return_fees.{{ $index }}.fee">
```

@joshhanley my next step would be converting this to an issue in Livewire. I'm a bit hesitant to open an issue with Laravel without a 100% certainty of this as they can be pretty harsh with closing issues. If it rightly belongs there, I'm happy to take it to that repo instead.

**Comment by ousid** [comment]
Have you used `wire:key=$key`? 

Take a look at [Troubleshooting](https://laravel-livewire.com/docs/2.x/troubleshooting#dom-diffing-issues) section on the docs

**Reply by benjibee** [reply]
@ousid thanks for the suggestion. I was attempting to document an issue with this specific usage of Livewire, though I know there are other ways to approach this as noted below.

**Comment by joshhanley** [comment, marked as answer]
@benjibee currently Livewire doesn't support binding to relations with multiple words, as you found on that other discussion. I've tried a few to times to get it working, but yeah as you described, there are issues with Laravel's `toArray` method and how we are serialising data.

Please don't convert this discussion to a Livewire issue, as we are no longer using issues, as we are now using discussions to track bugs and ideas, so this discussion is the place to talk about it 🙂
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

As to whether to open a Laravel issue, I don't think that is necessary (or wise as, as you suggested, they will probably just close it), mainly because they will not see any issue with how the `toArray` method works, as it has been that way for years.

I think it would be nice to come up with a Livewire solution for it, but I don't have the time to dig deep on this issue.

PR's are welcome though if you do come up with a way to get it to work 🙂

Personally, I've actually stopped using model binding all together and just either use primitive properties (string/ array) or use custom DTOs with the Wireable trait implemented. https://laravel-livewire.com/docs/2.x/properties#wireable-properties

Hope this helps!

**Reply by benjibee** [reply]
Thanks for getting back to me, and thanks for the clarifications. It might help to add a little note to the documentation as this sucked up a good chunk of time without any clear explanation.

If I come up with any good solution or fix I'll see if I'd have time for a PR. But you're suggested usage of primitives and wireable DTOs is likely the permanent fix for us, too. Though using binding is pretty ✨

**Reply by joshhanley** [reply]
@benjibee no worries! 🙂 yeah fair call, PR's are welcome to the docs https://github.com/livewire/docs

Ok thanks! Yeah simplifies a lot of things, specifically around the hydration/dehydration of properties. It is nice though to just be able to find to a few properties on a model and update them!

**Reply by wojciechczerniak** [reply]
@joshhanley @benjibee Can you share an example of a DTO and how it solves the issue with snake_case? You're creating a custom serializer for the entire model? How security is then handled (Livewire blocks properties that are not exposed through `$rules`)? Do you have custom validation rule for nested properties?

**Reply by joshhanley** [reply]
@wojciechczerniak the issue specifically relates to the Eloquent Collection/ Eloquent Model hydrator/serialiser, so using a DTO is performed different, and as such doesn't have the same issue.

The reason for that is with a DTO you need to implement the `Wireable` trait, which means you tell Livewire how to serialise and hydrate data yourself rather than Livewire trying to do it for you. https://laravel-livewire.com/docs/2.x/properties#wireable-properties

In the case of security, well you wouldn't make a DTO for all of the properties of in the entire model, you'd only do it for properties that the frontend needs access to. And for validation, no custom rules are needed, the validation should work on the DTO and its properties and nested properties without any issues.

To demonstrate all that, imaging we had a user with a `blogPosts` relation (which is a multiword relation that currently doesn't work with Livewire's serialisation):

```php
namespace App\DTOs;

use Livewire\Wireable;

class UserDTO implements Wireable
{
    public function __construct(
        public string $name,
        public string $email,
        public array $blogPosts = [],
    }
    
    public function toLivewire()
    {
        return [
            'name' => $this->name,
            'email' => $this->email,
            'blogPosts' => $this->blogPosts,
        ];
    }
    
    public function fromLivewire($value)
    {
        return new static(
            name: $value['name'],
            email: $value['email'],
            blogPosts: $value['blogPosts'],
        );
    }
}
```

Then to instantiate it:

```php
$user = User::with('blogPosts')->first();

$userDTO = new UserDTO(
    name: $user->name,
    email: $user->email,
    blogPosts: $user->blogPosts->toArray(),
);
```

As you can see we can definite how the data is serialised and sent to Livewire and re-hydrated, so we don't run into any issues.

Hope this helps!

**Reply by wojciechczerniak** [reply]
> And for validation, no custom rules are needed, the validation should work on the DTO and its properties and nested properties without any issues.

Cool, didn't knew that. 

Thank you, I will give it a try.

**Reply by joshhanley** [reply]
@wojciechczerniak no worries! 🙂

**Comment by Martleby** [comment]
I don't know if this is helpful or dangerous (or even possible in this scenario), but I encountered something similar in a non-Livewire context ages ago.

It turned out that you can actual set a public static property $snakeAttributes - provided by the model's hasAttributes trait - to false, which prevents relationships from being snake_cased on toArray().

As I recall I didn't encounter any problems, but after calling toArray I set it back to true again just to make sure.

If you use this and your house burns down, don't blame me!

**Reply by joshhanley** [reply]
@Martleby oh interesting! Hmm I wonder if we can make use of that, just depends where Livewire calls it (I can't remember exactly, it's been a while).

Oh but @Martleby this is OSS, so of course we will blame you if code changes we make to our own apps breaks everything 😆 /s
